File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []

        res = []

        stack = []
        stack.append(root)
        
        while len(stack) > 0:
            curr = stack.pop()

            if curr.right is not None:
                stack.append(curr.right)
        
            if curr.left is not None:
                stack.append(curr.left)

            res.append(curr.val)


        return res

File: Data Structures & Algorithms/same-binary-tree/submission-1.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
        const stack = [[p, q]];
        
        while (stack.length > 0) {
            const [rootOne, rootTwo] = stack.pop();

            if (!rootOne && !rootTwo) continue;
            if (!rootOne || !rootTwo || rootOne.val !== rootTwo.val) {
                return false;
            }

            stack.push([rootOne.left, rootTwo.left]);
            stack.push([rootOne.right, rootTwo.right]);
        }

        return true;
    }
}
